### PR TITLE
Replace addContentLengthHeader with ContentLength Middleware

### PR DESCRIPTION
**Pull request by akrabat**
This feature isn't needed in core and is better as optional middleware.

### CODE REVIEW

**Comment by damianopetrungaro** [comment]
@akrabat 
As discussed on Slack for the AppTest:



    public function testAddSettings()
    {
        $app = new App();
        $app->addSettings(['foo' => 'bar']);
        $this->assertAttributeContains('bar', 'settings', $app);
    }

    public function testAddSetting()
    {
        $app = new App();
        $app->addSetting('foo', 'bar');
        $this->assertAttributeContains('bar', 'settings', $app);
    }


with this it works 😄

**Comment by damianopetrungaro** [comment]
Or even better:

    public function testAddSettings()
    {
        $app = new App();
        $reflection = new ReflectionProperty(App::class, 'settings');
        $reflection->setAccessible(true);
        $reflection->setValue($app, []);
        $app->addSettings(['foo' => 'bar']);
        $this->assertSame(['foo' => 'bar'], $reflection->getValue($app));
    }

    public function testAddSetting()
    {
        $app = new App();
        $reflection = new ReflectionProperty(App::class, 'settings');
        $reflection->setAccessible(true);
        $reflection->setValue($app, []);
        $app->addSetting('foo', 'bar');
        $this->assertSame(['foo' => 'bar'], $reflection->getValue($app));
    }

**Comment by akrabat** [comment]
Ping @geggleto

File: Slim/Middleware/ContentLength.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+     * @return ResponseInterface                 PSR7 response
+     */
+    public function __invoke(ServerRequestInterface $request, ResponseInterface $response, callable $next)
+    {
+        $response = $next($request, $response);
+
+        // Add Content-Length header if not already added
+        $size = $response->getBody()->getSize();
+        if ($size !== null && !$response->hasHeader('Content-Length')) {
+            $response = $response->withHeader('Content-Length', (string) $size);
```
Comment (geggleto) [inline Slim/Middleware/ContentLength.php]: .... why are we casting it to a string? ... my gut says that is a really bad idea.

File: tests/Middleware/ContentLengthTest.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        $response = new Response();
+
+        $next = function (ServerRequestInterface $req, ResponseInterface $res) {
+            $res->write('Body');
+            return $res;
+        };
+
+        $newResponse = $mw($request, $response, $next);
+
+        $this->assertEquals(4, $newResponse->getHeaderLine('Content-Length'));
```
Comment (geggleto) [inline tests/Middleware/ContentLengthTest.php]: This test should also assert that the body's content is actually 'Body'

@akrabat what about multi-byte character sets?

**Review by geggleto — CHANGES_REQUESTED** [review]

File: Slim/Middleware/ContentLength.php (hunk 1 above)
Comment (akrabat) [reply Slim/Middleware/ContentLength.php]: [`withHeader()`](https://github.com/php-fig/http-message/blob/master/src/MessageInterface.php#L128) is defined as taking a string as the second parameter.

File: tests/Middleware/ContentLengthTest.php (hunk 2 above)
Comment (akrabat) [reply tests/Middleware/ContentLengthTest.php]: Why? The code under test doesn't touch the body and ensuring the body is the correct text is covered by tests on `Response` and `App`.

Comment (akrabat) [reply tests/Middleware/ContentLengthTest.php]: Size of multi-byte character sets is an issue for `Slim\Http\Stream()` and should be tested there I think?

i.e. I'm testing that the correct header is added as that's what this Middleware does.

Comment (akrabat) [reply tests/Middleware/ContentLengthTest.php]: There's an argument that the test should be:

```php
$this->assertEquals($response->getBody()->getSize(), $newResponse->getHeaderLine('Content-Length'));
```

or even:

```php
$this->assertTrue($newResponse->hasHeader('Content-Length'));
```

**Comment by akrabat** [comment]
Renamed to `ContentLengthMiddleware` as per https://github.com/slimphp/Slim/pull/2288#discussion_r133005546

File: UPGRADING.md [hunk 3]
```diff
@@ -1,9 +1,11 @@
 # How to upgrade
 
+* [2254] - You need to add the `Middleware\ContentLength` middleware if you want Slim to add this automatically.
```
Comment (campersau) [inline UPGRADING.md]: `ContentLengthMiddleware`?
Maybe the `OutputBuffering` should also be renamed?

Comment (akrabat) [reply UPGRADING.md]: Good catch. Thanks.

Yes, 'OutputBuffering' will need renaming too, but that's a separate PR if you fancy doing it?

**Review by geggleto — APPROVED** [review]

**Comment by akrabat** [comment]
Rebased to fix conflicts

**Pull request merged**